server/src/main/java/org/elasticsearch/index/mapper/MappingParser.java

```diff
@@ -117,8 +117,7 @@ private Mapping parse(String type, Map<String, Object> mapping) throws MapperPar
                 MetadataFieldMapper metadataFieldMapper = typeParser.parse(fieldName, fieldNodeMap, parserContext)
                     .build(MapperBuilderContext.ROOT);
                 metadataMappers.put(metadataFieldMapper.getClass(), metadataFieldMapper);
-                fieldNodeMap.remove("type");
-                checkNoRemainingFields(fieldName, fieldNodeMap);
+                assert fieldNodeMap.isEmpty();
```

Comment:
type was already removed within FieldMapper.Builder so it was always a no-op here. we already consume all the parameters so I think we can convert the check to an assertion here.